python/demo/demo_mixed-poisson.py

```diff
@@ -109,88 +104,296 @@
+# and (ii) bottom ($x_{1} = 0$) of the mesh/domain.
 
+# +
 fdim = msh.topology.dim - 1
-facets_top = mesh.locate_entities_boundary(msh, fdim, lambda x: np.isclose(x[1], 1.0))
-Q, _ = V0.collapse()
-dofs_top = fem.locate_dofs_topological((V0, Q), fdim, facets_top)
+dofs_top = fem.locate_dofs_topological(
+    V, fdim, mesh.locate_entities_boundary(msh, fdim, lambda x: np.isclose(x[1], 1.0))
+)
+dofs_bottom = fem.locate_dofs_topological(
+    V, fdim, mesh.locate_entities_boundary(msh, fdim, lambda x: np.isclose(x[1], 0.0))
+)
+# -
 
+# Now, we create Dirichlet boundary objects for the condition $\sigma
+# \cdot n = \sin(5 x_(0)$ on the top and bottom boundaries:
 
-def f1(x):
-    values = np.zeros((2, x.shape[1]))
-    values[1, :] = np.sin(5 * x[0])
-    return values
+# +
+f_h1 = fem.Function(V, dtype=dtype)
+f_h1.interpolate(lambda x: np.vstack((np.zeros_like(x[0]), np.sin(5 * x[0]))))
+bc_top = fem.dirichletbc(f_h1, dofs_top)
 
+f_h2 = fem.Function(V, dtype=dtype)
+f_h2.interpolate(lambda x: np.vstack((np.zeros_like(x[0]), -np.sin(5 * x[0]))))
+bc_bottom = fem.dirichletbc(f_h2, dofs_bottom)
 
-f_h1 = fem.Function(Q)
-f_h1.interpolate(f1)
-bc_top = fem.dirichletbc(f_h1, dofs_top, V0)
+bcs = [bc_top, bc_bottom]
+# -
 
+# Assemble the matrix operator `A` into a PETSc 'nested matrix', zero
+# rows and columns associated with a Dirichlet boundary condition and
+# placing 1 on the diagonal for Dirichlet constrained
+# degrees-of-freedom:
 
-facets_bottom = mesh.locate_entities_boundary(msh, fdim, lambda x: np.isclose(x[1], 0.0))
-dofs_bottom = fem.locate_dofs_topological((V0, Q), fdim, facets_bottom)
+# +
+A = fem.petsc.assemble_matrix_nest(a, bcs=bcs)
+A.assemble()
+# -
 
+# Assemble the RHS vector as a 'nested' vector and modify (apply
+# lifting) to account for the effect non-zero Dirichlet boundary
+# conditions. Then set Dirichlet boundary values in the RHS vector `b`:
 
-def f2(x):
-    values = np.zeros((2, x.shape[1]))
-    values[1, :] = -np.sin(5 * x[0])
-    return values
+# +
+b = fem.petsc.assemble_vector_nest(L)
+fem.petsc.apply_lifting_nest(b, a, bcs=bcs)
+for b_sub in b.getNestSubVecs():
+    b_sub.ghostUpdate(addv=PETSc.InsertMode.ADD, mode=PETSc.ScatterMode.REVERSE)
+
+bcs0 = fem.bcs_by_block(fem.extract_function_spaces(L), bcs)
+fem.petsc.set_bc_nest(b, bcs0)
+# -
+
+# Rather than solving the linear system $A x = b$, we will solve the
+# preconditioned problem $P^{-1} A x = P^{-1} b$. Commonly $P = A$, but this
+# does not lead to efficient solvers for saddle point problems.
+#
+# For this problem, we introduce the preconditioner
+# $$
+# a_p((\sigma, u), (\tau, v))
+# = \begin{bmatrix} \int_{\Omega} \sigma \cdot \tau + (\nabla \cdot
+# \sigma) (\nabla \cdot \tau) \ {\rm d} x  & 0 \\ 0 &
+# \int_{\Omega} u \cdot v \ {\rm d} x \end{bmatrix}
+# $$
+# and assemble it into the matrix `P`:
 
+# +
+a_p00 = inner(sigma, tau) * dx + inner(div(sigma), div(tau)) * dx
+a_p11 = inner(u, v) * dx
```

Comment:
Same comment as above wrt. extract_blocks. 
If we want to highlight both methods in 1 demo, we could keep this one as is as it is the simplest one.